server/src/main/java/com/linecorp/centraldogma/server/internal/api/ContentServiceV1.java

```diff
@@ -133,25 +134,22 @@ private static String rootDirIfEmpty(String path) {
                                      @RequestObject CommitMessageDto commitMessage,
                                      @RequestObject(ChangesRequestConverter.class)
                                              Iterable<Change<?>> changes) {
-        final CompletableFuture<Revision> normalizeFuture = repository.normalize(new Revision(revision));
+        final Revision normalizedRevision = repository.normalizeNow(new Revision(revision));
 
-        return normalizeFuture.thenCompose(normalizedRevision -> {
-            final CompletableFuture<Map<String, Change<?>>> changesFuture =
-                    repository.previewDiff(normalizedRevision, changes);
+        final CompletableFuture<Map<String, Change<?>>> changesFuture =
+                repository.previewDiff(normalizedRevision, changes);
 
+        return changesFuture.thenCompose(previewDiffs -> {
             final long commitTimeMillis = System.currentTimeMillis();
-            final CompletableFuture<Revision> resultRevisionFuture = changesFuture.thenCompose(
-                    previewDiffs -> {
-                        if (previewDiffs.size() == 0) {
-                            throw new RedundantChangeException("changes did not change anything: " + changes);
-                        }
-                        return push(commitTimeMillis, author, repository, normalizedRevision,
-                                    commitMessage, previewDiffs.values());
-                    });
+            if (previewDiffs.size() == 0) {
+                throw new RedundantChangeException("changes did not change anything: " + changes);
```

Comment:
This will generate a large string is `changes` is large. How about using the default constructor?